File: examples/python/customer/handlers/create_customer.py
"""Handler for CreateCustomer command."""

from datetime import datetime, timezone

import structlog
from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains

from handlers import CommandRejectedError


def handle_create_customer(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.CustomerState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle CreateCustomer command."""
    if state.name:
        raise CommandRejectedError("Customer already exists")

    cmd = domains.CreateCustomer()
    command_any.Unpack(cmd)

    if not cmd.name:
        raise CommandRejectedError("Customer name is required")
    if not cmd.email:
        raise CommandRejectedError("Customer email is required")

    log.info("creating_customer", name=cmd.name, email=cmd.email)

    event = domains.CustomerCreated(
        name=cmd.name,
        email=cmd.email,
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/rust/customer/src/handlers/create_customer.rs
//! Handler for CreateCustomer command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CreateCustomer, CustomerCreated, CustomerState};
use common::{decode_command, now, require_exists, require_not_exists, Result};

use crate::errmsg;
use crate::state::build_event_response;

/// Handle the CreateCustomer command.
///
/// Creates a new customer with the given name and email.
/// Fails if the customer already exists (non-empty name in state).
pub fn handle_create_customer(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &CustomerState,
    next_seq: u32,
) -> Result<EventBook> {
    require_not_exists(&state.name, errmsg::CUSTOMER_EXISTS)?;

    let cmd: CreateCustomer = decode_command(command_data)?;

    require_exists(&cmd.name, errmsg::NAME_REQUIRED)?;
    require_exists(&cmd.email, errmsg::EMAIL_REQUIRED)?;

    let event = CustomerCreated {
        name: cmd.name.clone(),
        email: cmd.email.clone(),
        created_at: Some(now()),
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.CustomerCreated",
        event,
    ))
}
